tools/src/wyvern/tools/parsing/coreparser/arch/ASTComponentTypeDecl.java

```diff
@@ -0,0 +1,175 @@
+                    if (d instanceof DefDeclaration) {
+                        List<FormalArg> args = ((DefDeclaration) d).getFormalArgs();
+                        if (args.size() != reqs.size()) {
+                            // inconsistent dependency errors
+                            ToolError.reportError(
+                                    ErrorMessage.COMPONENT_DEPENDENCY_INCONSISTENCY, this.getLocation(),
+                                    typeName);
+                        }
+                        for (FormalArg f : args) {
+                            String name = f.getName();
+                            String type = ((NominalType) f.getType()).getTypeMember();
+                            if (!reqs.get(name).equals(type)) {
+                                // inconsistent dependency error
+                                ToolError.reportError(
+                                        ErrorMessage.COMPONENT_DEPENDENCY_INCONSISTENCY, this.getLocation(),
+                                        typeName);
+                            }
+                        }
+                    }
+                }
+            }
+
+            /* Check that it exposes the correct fields */
+            List<Declaration> decls = ((New) expr).getDecls().stream()
+                    .filter(p -> p instanceof DefDeclaration)
+                    .collect(Collectors.toList());
+            for (Declaration d : decls) {
+                DefDeclaration dd = (DefDeclaration) d;
+                if (dd.getType() instanceof StructuralType) {
+                    List<DeclType> valdecltypes = ((StructuralType) (dd.getType()))
+                            .getDeclTypes().stream().filter(p -> p instanceof ValDeclType)
+                            .collect(Collectors.toList());
+
+                    for (DeclType dt : valdecltypes) {
+                        ValDeclType vdtype = (ValDeclType) dt;
+                        String name = vdtype.getName();
+                        String t = ((NominalType) vdtype.getSourceType().getValueType())
+                                .getTypeMember();
+                        if (provs.get(name) == null || !provs.get(name).equals(t)) {
+                            // inconsistent fields
+                            ToolError.reportError(
+                                    ErrorMessage.COMPONENT_DEPENDENCY_INCONSISTENCY, this.getLocation(),
+                                    typeName);
+                        } else {
+                            checkCount++;
+                        }
+                    }
+                    if (checkCount != valdecltypes.size()) {
+                        // inconsistent fields
+                        ToolError.reportError(
+                                ErrorMessage.COMPONENT_DEPENDENCY_INCONSISTENCY, this.getLocation(),
+                                typeName);
+                    }
+                }
+            }
```

Comment:
Lines 99-156 can be refactored a bit. Here's what I would suggest:

1. Get the DefDeclarations in expr as you do in lines 125-127.

2. Include an assertion statement to make sure that decls contains only one element.

3. Get a reference to the DefDeclaration in decls.

4. Perform the check in lines 103-119 along with the change that I have asked you to make.

5. Perform the check in lines 130-155.